test/tests_for_sdkmr/EE2Runjob_test.py

```diff
@@ -0,0 +1,268 @@
+
+    kafka.send_kafka_message.assert_any_call(KafkaCreateJob(user, job_id))
+    condor.run_job.assert_called_once_with(
+        params={
+            "method": method,
+            "source_ws_objects": [ws_obj1, ws_obj2],
+            "service_ver": git_commit,
+            "job_id": job_id,
+            "user_id": user,
+            "token": token,
+            "cg_resources_requirements": {
+                "client_group": client_group,
+                "request_cpus": cpus,
+                "request_memory": mem,
+            },
+        },
+        concierge_params=None,
+    )
+
+    # updated job data save
+    mongo.get_job.assert_called_once_with(job_id)
+
+    # update to queued state
+    got_job = sdkmr.save_job.call_args_list[1][0][0]
+    expected_job = Job()
+    expected_job.id = ObjectId(job_id)
+    expected_job.status = queued_state
+    # no way to test this really without code refactoring
+    expected_job.queued = got_job.queued
+
+    expected_job.scheduler_type = "condor"
+    expected_job.scheduler_id = cluster
+    assert_jobs_equal(got_job, expected_job)
+
+    kafka.send_kafka_message.assert_called_with(  # update to queued state
+        KafkaQueueChange(
+            job_id=job_id,
+            new_status=queued_state,
+            previous_status=created_state,
+            scheduler_id=cluster,
+        )
+    )
+    slack.run_job_message.assert_called_once_with(job_id, cluster, user)
+
+
+def assert_jobs_equal(got_job: Job, expected_job: Job):
+    """
+    Checks that the two jobs are equivalent, except that the 'updated' fields are checked that
+    they're within 1 second of each other.
+    """
+    # I could not get assert got_job == expected_job to ever work, or even
+    # JobRequirements1 == JobRequirements2 to work when their fields were identical.
+    # Also accessing the instance dict via vars() or __dict__ only produced {'_cls': 'Job'} so
+    # apparently MongoEngine does something very odd with the dict.
+    # Hence we do this disgusting hack instead. Note it will need to be updated any time a
+    # job field is added.
+
+    if not hasattr(got_job, "id"):
+        assert not hasattr(expected_job, "id")
+    else:
+        assert got_job.id == expected_job.id
+
+    # The Job class fills the updated field with the output of time.time on instantiation
+    # so we can't do a straight equality
+    assert abs(got_job.updated - expected_job.updated) < 1
+
+    job_fields = [
+        "user",
+        "authstrat",
+        "wsid",
+        "status",
+        "queued",
+        "estimating",
+        "running",
+        "finished",
+        "errormsg",
+        "msg",
+        "error",
+        "terminated_code",
+        "error_code",
+        "scheduler_type",
+        "scheduler_id",
+        "scheduler_estimator_id",
+        "job_output",
+        "condor_job_ads",
+        "child_jobs",
+        "batch_job",
+    ]
+
+    _super_hacky_equals(got_job, expected_job, job_fields)
```

Comment:
Can you rename this something more sensible? Initially I laughed at the name, but looking at the function, it's not actually hacky: you are just checking that a list of object attributes are equal.

Reply:
Here I disagree - this function is super hacky. If a field is changed in the `Job` class this function will stop working correctly, possibly silently. I definitely want future programmers to know this is a function to be avoided if possible.

Reply:
The `assert_job_equals` function is hacky, but `_super_hacky_equals` is not. You've already put in a disclaimer about why `assert_job_equals` is necessary and why it's horrible. Calling a function that uses a `for` loop and the `getattr` function  "hacky" is only going to invite confusion, since both are core to the functionality of python3.

Again, the reason that the function is required is hacky, which you have explained. The function itself is not. Please rename it to avoid confusion.

Reply:
>  Calling a function that uses a for loop and the getattr function "hacky" is only going to invite confusion, since both are core to the functionality of python3.

It's not hacky because of `getattr` or the `for` loop, it's hacky because it's implementing an equals function that is fragile and virtually guaranteed to break sometime in the future. No one should use that equals function unless there's no other alternative. The case that the name is appropriate is very strong here.

Reply:
Actually, I have an idea that may make us both happy. Give me a sec

Reply:
done